pkg/graveler/graveler.go

```diff
@@ -1427,25 +1444,36 @@ func (g *KVGraveler) Set(ctx context.Context, repository *RepositoryRecord, bran
 // didn't change while writing to the branch.
 func (g *KVGraveler) safeBranchWrite(ctx context.Context, log logging.Logger, repository *RepositoryRecord, branchID BranchID, stagingOperation func(branch *Branch) error) error {
 	var try int
+	// notFound is set if stagingOperation returns errNotFoundMustVerify
+	// that holds after verifying the staging token.  It causes
+	// safeBranchWrite to return ErrNotFound.
+	notFound := false
```

Comment:
Can we consider any data based error returned by the staging operation based on the cached information as something we like to verify? Or at least keep the original error and check for Graveler's NotFound as having a very specific error here can be very easy to miss when adding new functionality that can return not found.

Reply:
I think that verifying al errors will make things slow, see below.

I don't think that we should special-case `ErrNotFound` as you suggest, as it fails for _other_ possible returns of `ErrNotFound`.

I'd like to suggest something even more explicit: Would you be happy with changing the return signature of `stagingOperation` to be very explicit and Go-like:
```go
func(branch *Branch) (verify bool, err error)
```
where we only verify `if verify`?  (If you say "yes", please consider whether we should still verify when it returns `(false, nil)`; I believe that we should!)

### Why is it slow always to retry?

It is _incorrect_ to verify on all errors returned by `stagingOperation`.  Most errors returned will not be due to a branch lookup miss, and should not be retried.  Reverifying on every error makes every failed operation cost a GetBranch; most automated uses like to perform operations that can fail.

For instance, when you use lakeFSFS with the FileOutputCommitter, it repeatedly looks at the directory marker.  This is a very slow statObject call.  So we will probably want to change it to a "uploadObject if-not-exists" operation once we decide to improve its performance.  But if we retry on error, "uploadObject if-not-exists" again becomes an expensive operation!

Reply:
Suggest to use the same error wrapping technic but with a different error - ex: https://pkg.go.dev/github.com/cenkalti/backoff#Permanent

This will enable the code to identify the case of the specific error we like to verify, extract the underlaying error and not communicate the new error.
The code will be easy to read as the original error we wrap will be writtern at the same place we return it and not new kind of error.